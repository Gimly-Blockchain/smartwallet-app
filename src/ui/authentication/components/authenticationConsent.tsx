import React from 'react'
import { ButtonSection } from 'src/ui/structure/buttonSectionBottom'
import { Text, StyleSheet, View } from 'react-native'
import I18n from 'src/locales/i18n'
import strings from '../../../locales/strings'
import { Colors, Typography, Spacing } from 'src/styles'
import { IssuerCard } from '../../documents/components/issuerCard'
import { AuthenticationRequestSummary } from '../../../actions/sso/types'

interface Props {
  authenticationDetails: AuthenticationRequestSummary
  generateAndSendAuthenticationResponse: Function
  cancelAuthenticationRequest: Function
}

interface State {}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.backgroundLightMain,
  },
  topSection: {
    flex: 0.9,
  },
  issuerCard: {
    marginTop: Spacing.LG,
  },
  authRequestContainer: {
    flex: 1,
    paddingHorizontal: '10%',
    justifyContent: 'flex-start',
    alignItems: 'center',
    marginTop: Spacing.XL,
  },
  authRequestText: {
    ...Typography.baseFontStyles,
    ...Typography.centeredText,
    fontSize: Typography.textLG,
    marginTop: Spacing.SM,
  },
  buttonSection: {
    flex: 0.1,
  },
})

export class AuthenticationConsentComponent extends React.Component<
  Props,
  State
> {
  public state = {
    pending: false,
  }

  private handleConfirm = () => {
    this.setState({ pending: true })
    return this.props.generateAndSendAuthenticationResponse()
  }

  public render() {
    const { requester, description } = this.props.authenticationDetails
    return (
      <View style={styles.container}>
        <View style={styles.topSection}>
          <IssuerCard issuer={requester} style={styles.issuerCard} />
          <View style={styles.authRequestContainer}>
            <Text style={styles.authRequestText}>
              {I18n.t(strings.WOULD_YOU_LIKE_TO)}
            </Text>
            <Text
              style={[styles.authRequestText, { fontSize: Typography.text4XL }]}
            >
              {description}
            </Text>
            <Text style={styles.authRequestText}>
              {I18n.t(strings.WITH_YOUR_SMARTWALLET)}
            </Text>
          </View>
        </View>
        <View style={styles.buttonSection}>
          <ButtonSection
            disabled={this.state.pending}
            denyDisabled={this.state.pending}
            confirmText={I18n.t(strings.AUTHORIZE)}
            denyText={I18n.t(strings.DENY)}
            handleConfirm={this.handleConfirm}
            handleDeny={() => this.props.cancelAuthenticationRequest()}
          />
        </View>
      </View>
    )
  }
}
