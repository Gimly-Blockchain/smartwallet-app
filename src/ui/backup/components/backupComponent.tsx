import * as React from 'react'
import { ScrollView, StyleSheet, Text, View } from 'react-native'
import { Container } from '../../structure'
import { Colors } from '../../../styles'
import { fontMain } from '../../../styles/typography'
import { GradientButton } from '../../structure/gradientButton'
import { NavigationSection } from '../../structure/navigationSection'

interface Props {
  isLoading: boolean
  isAutoBackupEnabled: boolean
  enableAutoBackup: () => void
  onDisableAutoBackup: () => void
  exportBackup: () => void
  goBack: () => void
  lastBackup?: string
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.baseBlack,
    justifyContent: 'flex-start',
    alignItems: 'stretch',
  },
  scrollView: {
    padding: 18,
  },
  section: {
    backgroundColor: Colors.black,
    justifyContent: 'flex-start',
    alignItems: 'stretch',
    marginTop: 32,
    borderRadius: 8,
    padding: 24,
    paddingBottom: 28,
  },
  title: {
    fontFamily: fontMain,
    color: Colors.white,
    fontSize: 28,
  },
  subtitle: {
    fontFamily: fontMain,
    color: Colors.white,
    fontSize: 22,
    textAlign: 'center',
  },
  description: {
    color: Colors.white050,
    fontSize: 16,
    marginTop: 8,
    textAlign: 'center',
  },
  warning: {
    color: Colors.error,
    fontSize: 14,
    textAlign: 'center',
    marginTop: 28,
  },
  info: {
    textAlign: 'center',
    color: Colors.purpleMain,
    marginTop: 28,
  },
  buttonContainer: {
    marginTop: 22,
    width: '100%',
    paddingVertical: 9,
    margin: 0,
    borderRadius: 8,
  },
  buttonText: {
    fontSize: 18,
  },
})
const BackupComponent: React.FC<Props> = ({
  isLoading,
  enableAutoBackup,
  onDisableAutoBackup,
  exportBackup,
  isAutoBackupEnabled,
  goBack,
  lastBackup,
}) => {
  return (
    <Container style={styles.container}>
      <NavigationSection
        onNavigation={goBack}
        isBackButton={true}
        isDark={true}
      />
      <ScrollView style={styles.scrollView}>
        <Text style={styles.title}>Backup options</Text>

        <View style={styles.section}>
          <Text style={styles.subtitle}>Backup data on your own</Text>
          <Text style={styles.description}>
            Download an encrypted copy of the data in your SmartWallet to your
            device. Make sure to keep your backups up-to-date and stored
            somewhere safe
          </Text>
          <GradientButton
            containerStyle={styles.buttonContainer}
            textStyle={styles.buttonText}
            onPress={exportBackup}
            text={'Export file'}
            disabled={isLoading}
          />
          {lastBackup && (
            <Text style={styles.info}>
              {`Last backup ${new Date(lastBackup).toLocaleDateString()}`}
            </Text>
          )}
        </View>

        <View style={styles.section}>
          <Text style={styles.subtitle}>Jolocom backup service</Text>
          <Text style={styles.description}>
            Store your information using Jolocom’s backup service. With this
            option, the Jolocom team cannot view your data and your info stays
            automatically up-to-date.
          </Text>
          {isAutoBackupEnabled ? (
            <React.Fragment>
              <GradientButton
                containerStyle={styles.buttonContainer}
                textStyle={styles.buttonText}
                text={isLoading ? 'Deleting backup...' : 'Disable auto-backup'}
                onPress={onDisableAutoBackup}
                disabled={isLoading}
              />
              {lastBackup && (
                <Text style={styles.info}>
                  {`Last backup ${new Date(lastBackup).toLocaleDateString()}`}
                </Text>
              )}
            </React.Fragment>
          ) : (
            <React.Fragment>
              <GradientButton
                containerStyle={styles.buttonContainer}
                textStyle={styles.buttonText}
                onPress={enableAutoBackup}
                text={
                  isLoading ? 'Synchronization started…' : 'Use backup service'
                }
                disabled={isLoading}
              />
              <Text style={styles.warning}>
                By clicking here i agreed backing up my data on Jolocom’s
                central server and i know that it is not viewable by anyone
              </Text>
            </React.Fragment>
          )}
        </View>
      </ScrollView>
    </Container>
  )
}

export default BackupComponent
