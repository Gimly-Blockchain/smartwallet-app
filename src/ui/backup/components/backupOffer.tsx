import * as React from 'react'
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import { Container } from '../../structure'
import { Colors } from '../../../styles'
import { GradientButton } from '../../structure/gradientButton'
import { ActionSheet } from '../../structure/actionSheet'

interface Props {
  enableAutoBackup: () => void
  manualBackup: () => void
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.baseBlack,
    justifyContent: 'flex-start',
    alignItems: 'stretch',
  },
  title: {
    color: Colors.white080,
    fontSize: 22,
    textAlign: 'center',
    marginHorizontal: 40,
  },
  logoTitle: {
    color: Colors.white080,
    fontSize: 22,
    textAlign: 'center',
    marginHorizontal: 108,
  },
  bottomSheet: {
    position: 'absolute',
    bottom: 0,
    backgroundColor: Colors.black,
    height: 230,
    width: '100%',
    borderRadius: 22,
    borderBottomLeftRadius: 0,
    borderBottomRightRadius: 0,
    paddingTop: 28,
    paddingHorizontal: 20,
  },
  button: {
    marginTop: 28,
    marginHorizontal: 20,
  },
  info: {
    color: Colors.white050,
    fontSize: 14,
    textAlign: 'center',
    marginTop: 10,
  },
  textButton: {
    color: Colors.white,
    fontSize: 20,
    textAlign: 'center',
    marginTop: 20,
  },
})
export const BackupOfferComponent: React.FC<Props> = ({
  enableAutoBackup,
  manualBackup,
}) => {
  return (
    <Container style={styles.container}>
      <Text style={styles.title}>
        Store your data by yourself or try to use Jolocom backup service!
      </Text>
      <View>{/* TODO add correct icon*/}</View>
      <Text style={styles.logoTitle}>Automatic synchronization</Text>
      <ActionSheet showSlide={true}>
        <GradientButton
          onPress={enableAutoBackup}
          text={'Turn on backup service'}
        />
        <Text style={styles.info}>
          By clicking here i agreed backing up my data on Jolocomâ€™s central
          server server and i know that it is not viewable by anyone
        </Text>
        <TouchableOpacity onPress={manualBackup}>
          <Text style={styles.textButton}>Keep it manually</Text>
        </TouchableOpacity>
      </ActionSheet>
    </Container>
  )
}
