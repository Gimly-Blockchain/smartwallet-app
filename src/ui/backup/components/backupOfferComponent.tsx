import * as React from 'react'
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import { Container } from '../../structure'
import { Colors } from '../../../styles'
import { NavigationSection } from '../../structure/navigationSection'
import { GradientButton } from '../../structure/gradientButton'

interface Props {
  enableAutoBackup: () => void
  close: () => void
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.baseBlack,
    justifyContent: 'flex-start',
    alignItems: 'stretch',
  },
  title: {
    color: Colors.white080,
    fontSize: 22,
    textAlign: 'center',
    marginHorizontal: 40,
  },
  logoTitle: {
    color: Colors.white080,
    fontSize: 22,
    textAlign: 'center',
    marginHorizontal: 108,
  },
  bottomSheet: {
    position: 'absolute',
    bottom: 0,
    backgroundColor: Colors.black,
    height: 230,
    width: '100%',
    borderRadius: 22,
    borderBottomLeftRadius: 0,
    borderBottomRightRadius: 0,
    paddingTop: 28,
    paddingHorizontal: 20,
  },
  button: {
    marginTop: 28,
    marginHorizontal: 20,
  },
  info: {
    color: Colors.white050,
    fontSize: 14,
    textAlign: 'center',
    marginTop: 10,
  },
  textButton: {
    color: Colors.white,
    fontSize: 20,
    textAlign: 'center',
    marginTop: 20,
  },
})
const BackupOfferComponent: React.FC<Props> = ({ close, enableAutoBackup }) => {
  return (
    <Container style={styles.container}>
      <NavigationSection
        onNavigation={close}
        isBackButton={false}
        isDark={true}
      />
      <Text style={styles.title}>
        Store your data by yourself or try to use Jolocom backup service!
      </Text>
      <View>{/* TODO add correct icon*/}</View>
      <Text style={styles.logoTitle}>Automatic synchronization</Text>
      <View style={styles.bottomSheet}>
        <GradientButton
          onPress={enableAutoBackup}
          text={'Turn on backup service'}
        />
        <Text style={styles.info}>
          By clicking here i agreed backing up my data on Jolocomâ€™s central
          server and i know that it is not viewable by anyone
        </Text>
        {/*TODO add onPress*/}
        <TouchableOpacity>
          <Text style={styles.textButton}>Keep it manually</Text>
        </TouchableOpacity>
      </View>
    </Container>
  )
}

export default BackupOfferComponent
