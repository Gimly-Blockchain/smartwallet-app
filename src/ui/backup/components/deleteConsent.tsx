import { StyleSheet, Text, View } from 'react-native'
import { Colors } from '../../../styles'
import React from 'react'

const loaders = require('react-native-indicator')

interface Props {
  isLoading: boolean
  onCancel: () => void
  onAgree: () => void
}
const styles = StyleSheet.create({
  title: {
    color: Colors.white,
    fontSize: 26,
  },
  info: {
    color: Colors.white,
    fontSize: 14,
    marginTop: 8,
  },
  loaderContainer: {
    alignItems: 'center',
  },
  buttonSection: {
    marginTop: 20,
    flexDirection: 'row',
    justifyContent: 'flex-end',
  },
  button: {
    margin: 10,
    color: Colors.white,
    fontSize: 18,
  },
})
export const DeleteConsent: React.FC<Props> = ({
  isLoading,
  onCancel,
  onAgree,
}) => (
  <React.Fragment>
    <Text style={styles.title}>{'Disable auto-backup'}</Text>
    {isLoading ? (
      <View style={styles.loaderContainer}>
        <loaders.RippleLoader size={42} strokeWidth={2} color={Colors.beige} />
      </View>
    ) : (
      <React.Fragment>
        <Text style={styles.info}>
          {
            'By disabling this function you will erase all information from our server'
          }
        </Text>
        <View style={styles.buttonSection}>
          <Text style={styles.button} onPress={onCancel}>
            {'Cancel'}
          </Text>
          <Text style={styles.button} onPress={onAgree}>
            {'Agree'}
          </Text>
        </View>
      </React.Fragment>
    )}
  </React.Fragment>
)
