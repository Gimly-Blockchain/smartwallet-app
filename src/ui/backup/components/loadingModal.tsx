import React from 'react'
import { StyleSheet, View } from 'react-native'
import { Colors } from '../../../styles'
import Modal from 'react-native-modal'

interface Props {
  isVisible: boolean
  close?: () => void
}

const styles = StyleSheet.create({
  container: {
    borderRadius: 8,
    backgroundColor: '#444145',
    justifyContent: 'center',
    alignItems: 'stretch',
    padding: 24,
  },
  title: {
    color: Colors.white,
    fontSize: 26,
    textAlign: 'center',
  },
  loaderContainer: {
    margin: 24,
  },
  waring: {
    color: Colors.error,
    fontSize: 14,
    textAlign: 'center',
  },
})
export const LoadingModal: React.FC<Props> = ({
  isVisible,
  close,
  children,
}) => (
  <Modal
    isVisible={isVisible}
    animationIn={'fadeIn'}
    animationOut={'fadeOut'}
    onBackdropPress={close}
  >
    <View style={styles.container}>{children}</View>
  </Modal>
)
