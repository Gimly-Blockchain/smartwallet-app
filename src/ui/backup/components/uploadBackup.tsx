import { StyleSheet, Text, View } from 'react-native'
import { Colors } from '../../../styles'
import React from 'react'
const loaders = require('react-native-indicator')

const styles = StyleSheet.create({
  title: {
    color: Colors.white,
    fontSize: 26,
    textAlign: 'center',
  },
  loaderContainer: {
    alignItems: 'center',
    margin: 24,
  },
  waring: {
    color: Colors.error,
    fontSize: 14,
    textAlign: 'center',
  },
})
export const UploadBackup = () => (
  <React.Fragment>
    <Text style={styles.title}>{'Backing-up with Jolocom'}</Text>
    <View style={styles.loaderContainer}>
      <loaders.RippleLoader size={42} strokeWidth={2} color={Colors.beige} />
    </View>
    <Text style={styles.waring}>
      {
        'The process of secure synchronization of current data and configuration of auto-update function has started'
      }
    </Text>
  </React.Fragment>
)
