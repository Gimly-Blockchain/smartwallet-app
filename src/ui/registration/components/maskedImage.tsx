import React from 'react'
import {
  PanResponder,
  PanResponderInstance,
  GestureResponderEvent,
} from 'react-native'
import { Svg, Path, Circle } from 'react-native-svg'
import { JolocomTheme } from 'src/styles/jolocom-theme'

interface Props {
  addPoint: (x: number, y: number) => void
}

interface State {
  linesPts: number[]
  prevX: number
  prevY: number
  curX: number
  curY: number
  pathD: string
  circles: number[][]
}

const MIN_DISTANCE_SQ = 50
const MAX_LINE_PTS = 150

export class MaskedImageComponent extends React.Component<Props, State> {
  private panResponder!: PanResponderInstance

  state: State = {
    linesPts: [],
    prevX: 0,
    prevY: 0,
    curX: 0,
    curY: 0,
    pathD: '',
    circles: [],
  }

  componentWillMount() {
    this.panResponder = this.getConfiguredPanResponder()
  }

  private getConfiguredPanResponder(): PanResponderInstance {
    return PanResponder.create({
      onMoveShouldSetPanResponder: () => true,
      onMoveShouldSetPanResponderCapture: () => true,
      onPanResponderGrant: this.handleDrawStart,
      onPanResponderMove: this.handleDraw,
    })
  }

  private handleDrawStart = (e: GestureResponderEvent): void => {
    const { locationX: x, locationY: y } = e.nativeEvent
    this.props.addPoint(x, y)

    this.setState({
      curX: x,
      curY: y,
      prevX: x,
      prevY: y,
      pathD: this.state.pathD + `M${x},${y} `,
    })
  }

  private handleDraw = (e: GestureResponderEvent): void => {
    const { locationX: curX, locationY: curY } = e.nativeEvent
    const { prevX, prevY, linesPts, circles, pathD } = this.state

    this.props.addPoint(curX, curY)
    const dist_sq = Math.abs(curX - prevX) + Math.abs(curY - prevY)

    if (dist_sq > MIN_DISTANCE_SQ) {
      for (let i = 0; i < linesPts.length; i += 4) {
        const intsct = this.getIntersection(
          linesPts[i],
          linesPts[i + 1],
          linesPts[i + 2],
          linesPts[i + 3],
          prevX,
          prevY,
          curX,
          curY,
        )
        if (intsct) circles.push(intsct)
      }
      linesPts.push(prevX, prevY, curX, curY)
      if (linesPts.length > MAX_LINE_PTS) linesPts.splice(0, 4)
      this.setState({
        curX,
        curY,
        prevX: curX,
        prevY: curY,
        pathD: pathD + `L${curX},${curY} `,
        linesPts,
        circles,
      })
    } else {
      this.setState({
        curX,
        curY,
      })
    }
  }

  // returns [x, y] for the intersection of (a,b)->(c,d) with (p,q)->(r,s),
  // or undefined if they don't intersect
  private getIntersection(
    a: number,
    b: number,
    c: number,
    d: number,
    p: number,
    q: number,
    r: number,
    s: number,
  ): number[] | undefined {
    let det, gamma, lambda
    det = (c - a) * (s - q) - (r - p) * (d - b)
    if (det === 0) return

    lambda = ((s - q) * (r - a) + (p - r) * (s - b)) / det
    gamma = ((b - d) * (r - a) + (c - a) * (s - b)) / det
    if (0 < lambda && lambda < 1 && (0 < gamma && gamma < 1)) {
      const x = a + lambda * (c - a)
      const y = b + lambda * (d - b)
      return [x, y]
    }
    return
  }

  render() {
    return (
      <Svg width="100%" height="100%" {...this.panResponder.panHandlers}>
        <Path
          d={this.state.pathD + `L${this.state.curX},${this.state.curY}`}
          fill="none"
          stroke={JolocomTheme.primaryColorSandInactive}
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeDasharray="1,4%"
          strokeWidth="1%"
        />
        {this.state.circles.map((c, i) => (
          <Circle
            key={i}
            cx={c[0]}
            cy={c[1]}
            r={5}
            fill={JolocomTheme.primaryColorSand}
          />
        ))}
      </Svg>
    )
  }
}
