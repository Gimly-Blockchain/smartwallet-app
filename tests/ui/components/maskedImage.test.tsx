import React from 'react'
import { shallow } from 'enzyme'
import { MaskedImageComponent } from 'src/ui/registration/components/maskedImage'

describe('MaskedImage component', () => {
  const props = {
    addPoint: (x: number, y: number) => null,
    disabled: false,
  }

  it('correctly calls the function to configure PanResponder on component mount', () => {
    const configureSpy = jest.spyOn(
      MaskedImageComponent.prototype,
      // @ts-ignore because getConfiguredPanResponder is private
      'getConfiguredPanResponder',
    )
    shallow(<MaskedImageComponent {...props} />)
    expect(configureSpy).toHaveBeenCalledTimes(1)
  })

  it('matches the snapshot with empty current path string', () => {
    const rendered = shallow<MaskedImageComponent>(<MaskedImageComponent {...props} />)
    expect(rendered).toMatchSnapshot()
    expect(rendered.state().pathDs[0]).toEqual('')
  })

  it('correctly handles a gesture start', () => {
    const x = 191,
      y = 381,
      mockNativeEvent = {
        nativeEvent: {
          locationX: x,
          locationY: y,
        },
      }
    const rendered = shallow<MaskedImageComponent>(<MaskedImageComponent {...props} />)
    const instance = rendered.instance()

    // @ts-ignore because handleDrawStart is private
    instance.handleDrawStart(mockNativeEvent)
    expect(rendered.state().pathDs[0]).toEqual(`M${x},${y}`)
  })

  it('correctly handles gesture continuation', () => {
    const x = 188,
      y = 381,
      mockNativeEvent = {
        nativeEvent: {
          locationX: x,
          locationY: y,
        },
      }

    const rendered = shallow<MaskedImageComponent>(<MaskedImageComponent {...props} />)
    const instance = rendered.instance()

    // @ts-ignore because handleDraw is private
    instance.handleDraw(mockNativeEvent)
    expect(rendered.state().pathDs[0]).toEqual(`L${x},${y}`)
  })
})
