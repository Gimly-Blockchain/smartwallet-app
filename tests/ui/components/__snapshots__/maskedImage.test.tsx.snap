// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaskedImage component matches the snapshot with empty current path string 1`] = `
<Svg
  height="100%"
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  preserveAspectRatio="xMidYMid meet"
  width="100%"
>
  <Path
    d=""
    fill="none"
    stroke="rgba(255, 239, 223, 0.38)"
    strokeDasharray="1,4%"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth="1%"
  />
</Svg>
`;
